rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,126 @@
+
+> Note: Some lit-html tests need to be skipped, for example dev mode warnings that will no longer execute if the template has been compiled.
+
+### Performance and Code Size Impact
+
+This RFC has been prototyped: https://github.com/lit/lit/pull/3984 to gather benchmarks. Note that these are all subject to change based on the final implementation. These initial benchmarks support that compiling templates do result in a measurable performance improvement, with greater performance improvements occurring in template heavy contexts.
+
+**Kitchen sink** average tachometer benchmark results:
+
+- [**6.7% faster**]: `render-compiled`: 7.93ms - 8.10ms, `render-uncompiled`: 8.50ms - 8.69ms
+- [roughly same]: `update-compiled`: 17.92ms - 18.56ms, `update-uncompiled`: 18.08ms - 19.00ms
+- [roughly same]: `nop-update-compiled`: 5.53ms - 5.66ms, `nop-update-uncompiled`: 5.64ms - 5.80ms
+ - [**9.3% larger gzipped file size**]: `index.js` uncompiled is 2216 bytes gzipped, or compiled it is 2443 bytes gzipped.
+
+**Template Heavy** average tachometer benchmark results:
+
+ - [**46% faster**]: `render-compiled`: 9.51ms - 9.67ms, `render-uncompiled`: 17.66ms - 18.03ms
+ - [**21% faster**]: `update-compiled`: 26.40 - 26.91ms, `update-uncompiled`: 33.62 - 34.51ms
+ - [**5.6% larger gzipped file size**]: `template-heavy.js` uncompiled is 202,544 bytes gzipped, or compiled it is 214,495 bytes gzipped.
+
+**repeat** average tachometer benchmark results:
+
+ - [**14% faster**]: `render-compiled`: 7.40ms - 7.62ms, `render-uncompiled`: 8.56ms - 8.94ms
+ - [**8% faster**]: `update-compiled`: 189.57ms - 197.54ms, `update-uncompiled`: 206.51ms - 218.68ms
+ - [**7.1% larger gzipped file size**]: `index.js` uncompiled is 3072 bytes gzipped, or compiled it is 3307 bytes gzipped.
+
+### Interoperability
+
+This labs package is not interoperable as it specifically optimizes Lit template results using specific knowledge of the lit-html internal render phases.
+
+### Security Impact
+
+The prepared compiled template must be a tagged template literal to defend against JSON injection attacks. Otherwise the `@lit-labs/compiler` package is only a transform and has no security impact.
+
+### Documentation Plan
+
+This package will initially be documented in its own README. If it stays on track to graduation, we should document this package under the *Tools and Workflows* section on lit.dev. The package on GitHub will also include a TypeScript and JavaScript example of using `@lit-labs/compiler`.
+
+## Downsides
+
+The downside of this approach is that it requires the code to have a build step which can add complexity and build specific setups. There are also slight mismatches in the precompiled prepared HTML, such as lit markers which are not required with pre-compilation.
+
+## Alternatives
+
+No alternatives were considered and rejected.
```

Comment:
We could not release this. It will have an initial implementation cost, and an ongoing maintenance cost, but the performance wins are large enough that it seems worth it. We also expect this code to need only occasional maintenance from TypeScript updates, as Lit's template result API has been very stable

Reply:
Much better! I hope you don't mind that I took your words.